storey/flow.py

```diff
@@ -386,6 +388,104 @@ async def _do(self, event):
                 await self._worker_awaitable
 
 
+class _PendingEvent:
+    def __init__(self):
+        self.in_flight = []
+        self.pending = []
+
+
+class _ConcurrentByKeyJobExecution(Flow):
+    def __init__(self, max_in_flight=8, **kwargs):
+        Flow.__init__(self, **kwargs)
+        self._max_in_flight = max_in_flight
+        self._q = None
+        self._pending_by_key = {}
+
+    async def _worker(self):
+        try:
+            while True:
+                job = await self._q.get()
+                if job is _termination_obj:
+                    for key, pending_event in self._pending_by_key.items():
+                        if self._pending_by_key[key].pending and not self._pending_by_key[key].in_flight:
+                            resp = await self._process_event(pending_event.pending[0])
+                            for event in pending_event.pending:
+                                await self._handle_completed(event, resp)
+                    if self._q.empty():
+                        break
+                    else:
+                        await self._q.put(_termination_obj)
+                        continue
+
+                event = job[0]
+                completed = await job[1]
+
+                for event in self._pending_by_key[event.key].in_flight:
+                    await self._handle_completed(event, completed)
+                self._pending_by_key[event.key].in_flight = []
+
+                # If we got more pending events for the same key process them
+                if self._pending_by_key[event.key].pending:
+                    first_event = self._pending_by_key[event.key].pending[0]
+                    self._pending_by_key[event.key].in_flight = self._pending_by_key[event.key].pending
+                    self._pending_by_key[event.key].pending = []
+
+                    task = self._process_event(first_event)
```

Comment:
Don't this code only process the first pending event? But it moves all the events from pending to in_flight?

Reply:
well for the current implementation it's fine since the data i actually want to save is in the cache object.
But maybe for other implementations, other drivers might want the raw events in addition to the cached data.. 🤷‍♂️